src/Repo_new.java

```diff
@@ -0,0 +1,804 @@
+
+    private static String getFileName(final String path) {
+        final int slash = path.lastIndexOf('/');
+
+        if (slash == -1) {
+            return path;
+        } else if (slash == (path.length() - 1)) {
+            return "";
+        } else {
+            return path.substring(slash + 1);
+        }
+    }
+
+    private Folder getOrCreateParentFolder(final String path) {
+        final int slash = path.lastIndexOf('/');
+
+        if (slash == -1) {
+            return getRootFolder();
+        } else {
+            return getOrCreateFolder(path.substring(0, slash));
+        }
+    }
+
+    private static File createFile(final File dir, final String path) {
+        File f = new File(dir, path);
+        for (int i = 1; f.exists(); i++) {
+            f = new File(dir, path + i);
+        }
+        return f;
+    }
+
+    private static String getViewName(final StarTeamURL url) {
+        final String src = url.getSource();
+        final String proj = '/' + url.getProjectName() + '/';
+        int ind = src.indexOf(proj);
+
+        if (ind != -1) {
+            String view = src.substring(ind + proj.length());
+
+            ind = view.indexOf('/');
+
+            if (ind != -1) {
+                view = view.substring(0, ind);
+            }
+
+            return view;
+        }
+
+        return "main";
+    }
+
+    private static Folder getRootFolder(final StarTeamURL url, final View v) {
+        Folder f = v.getRootFolder();
+        boolean skipViewName = true;
+
+        stLoop: for (final StringTokenizer st = url.getFolders(); st
+                .hasMoreTokens();) {
+            final String name = st.nextToken();
+
+            if (skipViewName) {
+                skipViewName = false;
+                if (name.equals(v.getName())) {
+                    continue;
+                }
+            }
+
+            for (final Folder sub : f.getSubFolders()) {
+                if (name.equals(sub.getName())) {
+                    f = sub;
+                    continue stLoop;
+                }
+            }
+
+            throw new IllegalArgumentException("Invalid url '"
+                    + url.getSource() + "': folder " + name
+                    + " does not exist.");
+        }
+
+        return f;
+    }
+
+    private int autoLogOn(final Server server) {
+        try {
+            // Without synchronized autoLogOn() sometimes fails on concurrent
+            // access.
+            synchronized (Repo.class) {
+                return server.autoLogOn();
+            }
+        } catch (final Throwable ex) {
+            if (_logger.isDebugEnabled()) {
+                _logger.error(ex.getMessage(), ex);
+            }
+            return 0;
+        }
+    }
+
+    private static void cacheLogOnCredentials(final Server server,
+            final String userName, final String password) {
+        try {
+            server.cacheLogOnCredentials(userName, password);
+        } catch (final Throwable ex) {
```

Comment:
![BLOCKER](https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER') Catch Exception instead of Throwable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=squid%3AS1181)
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=squid%3AS1166)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either remove or fill this block of code. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://localhost:9000/coding_rules#rule_key=squid%3AS00108)